pkg/backend/net/net.go

```diff
@@ -194,21 +194,24 @@ func (b Backend) do(ctx context.Context, trace types.Trace, req *http.Request) (
 	t0 := time.Now()
 	resp, err := b.client.Do(req)
 	trace.AddHTTPCall(t0)
-	if err != nil {
-		if resp != nil && resp.Body != nil {
-			resp.Body.Close()
-		}
-		return "", nil, err
+
+	var body []byte
+	var bodyErr error
+	if resp != nil && resp.Body != nil {
+		t1 := time.Now()
+		body, bodyErr = ioutil.ReadAll(resp.Body)
+		resp.Body.Close()
+		trace.AddReadBody(t1)
 	}
 
-	t1 := time.Now()
-	body, err := ioutil.ReadAll(resp.Body)
-	resp.Body.Close()
-	trace.AddReadBody(t1)
 	if err != nil {
```

Comment:
[minor] I would suggest moving this check as early as 10 lines up

Reply:
... before the [`client.Do`](https://github.com/bookingcom/carbonapi/blob/67b6cce823718b6272c15fc1becc5247d39222ec/pkg/backend/net/net.go#L195) call? I think I misunderstood something here.

Reply:
I meant: right after line 197: trace.AddHTTPCall(t0). [This is the revision](https://github.com/bookingcom/carbonapi/blob/67b6cce823718b6272c15fc1becc5247d39222ec/pkg/backend/net/net.go) I'm checking.

Reply:
I want to believe you are right, but I think I need to either go read some stdlib code or bang out some examples to figure out what's going on.

The `net/http` documentation does say:

> It is the caller's responsibility to close Body. The default HTTP client's Transport may not reuse HTTP/1.x "keep-alive" TCP connections if the Body is not read to completion and closed.

It also does **not** say whether when we write
```
resp, err := client.Do(req)
```
then `resp` and `err` can both be non-nil, or when what would happen. If that can happen, then we'd need to close `resp.Body` before returning from the error.

There are [some sources](http://devs.cloudimmunity.com/gotchas-and-common-mistakes-in-go-golang/) out there who claim this can happen:

> Most of the time when your http request fails the resp variable will be nil and the err variable will be non-nil. However, when you get a redirection failure both variables will be non-nil. This means you can still end up with a leak.

The code I wrote is the most paranoid reaction to the `net/http` docs and the statements in that article. It **will** read all of `resp.Body` and close it if it's ever non-nil.

Reply:
"On error, any Response can be ignored." - https://golang.org/pkg/net/http/#Client.Do
Also this [stackoverflow answer](https://stackoverflow.com/questions/33238518/what-could-happen-if-i-dont-close-response-body-in-golang) talks about this.

Reply:
Oh, I was not aware of this case. Thanks a lot for clarifying. Your original code now makes a perfect sense to me.

Reply:
Ooh, that's a good catch! I can think of a few places where this should be checked in other codebases...